File: code-challenges/401/linked-lists/linked-list.js
'use strict';

Features:

//Create a Node class that has properties for the value stored in the Node, and a pointer to the next Node.
// Within your LinkedList class, include a head property. Upon instantiation, an empty Linked List should be created.
// Define a method called insert which takes any value as an argument and adds a new node with that value to the head of the list with an O(1) Time performance.
// Define a method called includes which takes any value as an argument and returns a boolean result depending on whether that value exists as a Node’s value somewhere within the list.
// Define a method called toString (or __str__ in Python) which takes in no arguments and returns a string representing all the values in the Linked List, formatted as:
// "{ a } -> { b } -> { c } -> NULL"
// Any exceptions or errors that come from your code should be semantic, capturable errors. For example, rather than a default error thrown by your language, your code should raise/throw a custom, semantic error that describes what went wrong in calling the methods you wrote for this lab.
// Be sure to follow your language/frameworks standard naming conventions (e.g. C# uses PascalCasing for all method and class names).
// Structure and Testing
// Utilize the Single-responsibility principle: any methods you write should be clean, reusable, abstract component parts to the whole challenge. You will be given feedback and marked down if you attempt to define a large, complex algorithm in one function definition.

// Write tests to prove the following functionality:

// Can successfully instantiate an empty linked list
// Can properly insert into the linked list
// The head property will properly point to the first node in the linked list
// Can properly insert multiple nodes into the linked list
// Will return true when finding a value within the linked list that exists
// Will return false when searching for a value in the linked list that does not exist
// Can properly return a collection of all the values that exist in the linked list
// Ensure your tests are passing before you submit your solution.



// Code challenge 6

// Feature Tasks
// Write the following methods for the Linked List class:

// .append(value) which adds a new node with the given value to the end of the list
// .insertBefore(value, newVal) which add a new node with the given newValue immediately before the first value node
// .insertAfter(value, newVal) which add a new node with the given newValue immediately after the first value node
// Examples
// .append(value)
// Input	Args	Output
// head -> [1] -> [3] -> [2] -> X	5	head -> [1] -> [3] -> [2] -> [5] -> X
// head -> X	1	head -> [1] -> X
// .insertBefore(value, newVal)
// Input	Args	Output
// head -> [1] -> [3] -> [2] -> X	3, 5	head -> [1] -> [5] -> [3] -> [2] -> X
// head -> [1] -> [3] -> [2] -> X	1, 5	head -> [5] -> [1] -> [3] -> [2] -> X
// head -> [1] -> [2] -> [2] -> X	2, 5	head -> [1] -> [5] -> [2] -> [2] -> X
// head -> [1] -> [3] -> [2] -> X	4, 5	Exception
// .insertAfter(value, newVal)
// Input	Args	Output
// head -> [1] -> [3] -> [2] -> X	3, 5	head -> [1] -> [3] -> [5] -> [2] -> X
// head -> [1] -> [3] -> [2] -> X	2, 5	head -> [1] -> [3] -> [2] -> [5] -> X
// head -> [1] -> [2] -> [2] -> X	2, 5	head -> [1] -> [2] -> [5] -> [2] -> X
// head -> [1] -> [3] -> [2] -> X	4, 5	Exception
// Unit Tests
// Utilize the Single-responsibility principle: any methods you write should be clean, reusable, abstract component parts to the whole challenge. You will be given feedback and marked down if you attempt to define a large, complex algorithm in one function definition.

// You have access to the Node class and all the properties on the Linked List class.

// Write tests to prove the following functionality:

// Can successfully add a node to the end of the linked list
// Can successfully add multiple nodes to the end of a linked list
// Can successfully insert a node before a node located i the middle of a linked list
// Can successfully insert a node before the first node of a linked list
// Can successfully insert after a node in the middle of the linked list
// Can successfully insert a node after the last node of the linked list
// Unit tests must be passing before you submit your final solution code.

// Stretch Goal
// Once you’ve achieved a working solution, write an additional method to delete a node with the given value from the linked list.

File: Python/linked_list/linked_list.py
# Create a Node class that has properties for the value stored in the Node, and a pointer to the next Node.
class Node:
    def __init__(self, data, next):
        self.data = data
        self.next = None

    def __repr__(self):
        return self.data

# Within your LinkedList class, include a head property. Upon instantiation, an empty Linked List should be created.
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def __repr__(self):
        node = self.head
        nodes = []
        while node is not None:
            nodes.append(node.data)
            node = node.next
        nodes.append("None")
        return " -> ".join(nodes)

# Define a method called toString (or __str__ in Python) which takes in no arguments and returns a string representing all the values in the Linked List, formatted as:
# "{ a } -> { b } -> { c } -> NULL"
    def __str__(self):
        elements = []
        current = self.head
        while current.next != None:
            current = current.next
            elements.append(current.data)
        print(''.join("{} -> ".format(*k) for k in enumerate(elements))+'NULL')

# Define a method called insert which takes any value as an argument and adds a new node with that value to the head of the list with an O(1) Time performance.
    def insert(self, data):
        old_head = self.head
        self.head = Node(data, self.head)
        self.head.next = old_head

# Define a method called includes which takes any value as an argument and returns a boolean result depending on whether that value exists as a Node’s value somewhere within the list.
    def includes(self, data):
        if self.head == None:
            return False
        current_node = self.head
        while current_node.next != None:
            current_node = current_node.next
            if current_node.data == data:
                return True
        return False

# returns all values in Linked_list
    def return_all(self):
        all_values = []
        current_node = self.head
        while current_node != None:
            all_values.append(current_node.data)
            print(all_values)
            current_node = current_node.next
        return all_values

# .append(value) which adds a new node with the given value to the end of the list
    # def append(self, data)
    def append_node(self, data):
        newNode = Node(data, next=None)
        if(self.head):
            current = self.head
            while(current.next):
                current = current.next
            current.next = newNode
        else:
            self.head = newNode

# .insertBefore(value, newVal) which add a new node with the given newValue immediately before the first value node

    def insert_before(self, data, newData):
        newNode = Node(data, next=None)
        current_node = self.head
        if current_node == None:
            return "Sorry no nodes present"
        else:
            found_node = None
            while current_node:
                if current_node.next == data:
                    found_node = True
                    before_insert = current_node
                    after_insert = current_node.next
                    before_insert.next = newNode
                    newNode.next = after_insert
                    current_node = current_node.next
                else:
                    current_node = current_node.next
            if found_node != True:
                return "Could not find node"


# .insertAfter(value, newVal) which add a new node with the given newValue immediately after the first value node

    # def insert_after(data, newData)

# only run this when running directly as a "script"
if __name__ == "__main__":
    ll=LinkedList()
    ll.insert(1)
    ll.insert(2)
    ll.insert(3)
    ll.insert(4)
    ll.return_all()

